riscv/src/elf.rs

```diff
@@ -0,0 +1,1229 @@
+        entry_point: elf.entry as u32,
+    }
+}
+
+fn pie_relocate_data_sections(
+    elf: &Elf,
+    address_map: &AddressMap,
+    data_map: &mut BTreeMap<u32, Data>,
+    referenced_text_addrs: &mut BTreeSet<u32>,
+) {
+    // In PIE files, we can read the dynamic relocation table.
+    for r in elf.dynrelas.iter() {
+        let addr = r.r_offset as u32;
+        if !address_map.is_in_data_section(addr) {
+            unimplemented!("We assumed all dynamic relocations were data relocations!");
+        }
+
+        // We only support the R_RISCV_RELATIVE relocation type:
+        assert_eq!(r.r_type, R_RISCV_RELATIVE, "Unsupported relocation type!");
+
+        let data_value = r.r_addend.unwrap() as u32;
+
+        if address_map.is_in_text_section(data_value) {
+            data_map.insert(addr, Data::TextLabel(data_value));
+
+            // We also need to add the referenced address to the list of text
+            // addresses, so we can generate the label.
+            referenced_text_addrs.insert(data_value);
+        } else {
+            data_map.insert(addr, Data::Value(data_value));
+        }
+    }
+
+    assert_eq!(elf.dynrels.len(), 0, "Unsupported relocation type!");
+}
+
+fn static_relocate_data_sections(
+    elf: &Elf,
+    address_map: &AddressMap,
+    data_map: &mut BTreeMap<u32, Data>,
+    referenced_text_addrs: &mut BTreeSet<u32>,
+) {
+    // In non-PIE files, we need to use the linking relocation table.
+    for r in elf.shdr_relocs.iter().flat_map(|(_, relocs)| relocs.iter()) {
+        let addr = r.r_offset as u32;
+        if !address_map.is_in_data_section(addr) {
+            // Relocation of the text section has already been handled in instruction lifting.
+            continue;
+        }
+
+        // We only support the R_RISCV_32 relocation type for the data section:
+        assert_eq!(r.r_type, R_RISCV_32, "Unsupported relocation type!");
+
+        let Entry::Occupied(mut entry) = data_map.entry(r.r_offset as u32) else {
+            panic!("Unexpected 0 in relocated data entry!");
+        };
+
+        let Data::Value(original_addr) = *entry.get() else {
+            panic!("Related entry already replaced with a label!");
+        };
+
+        if address_map.is_in_text_section(original_addr) {
+            entry.insert(Data::TextLabel(original_addr));
+
+            // We also need to add the referenced address to the list of text
+            // addresses, so we can generate the label.
+            referenced_text_addrs.insert(original_addr);
+        }
+    }
+}
+
+/// Index the symbols by their addresses.
+struct SymbolTable(HashMap<u32, String>);
+
+impl SymbolTable {
+    fn new(elf: &Elf) -> SymbolTable {
+        // TODO: read the symbols from the debug information to be more comprehensive.
+        let mut deduplicator = HashMap::new();
+        for sym in elf.syms.iter() {
+            if sym.st_name == 0 || (sym.st_type() != STT_OBJECT && sym.st_type() != STT_FUNC) {
```

Comment:
can you add a comment about what this if is filtering?

Reply:
done